imperial.js

```diff
@@ -1,20 +1,62 @@
-import { Nation } from "./constants.js";
+import { Nation, Strings, NationCard } from "./constants.js";
 import Action from "./action.js";
 
+Map.fromIterable = (iter, getValue) => {
+  const m = new Map();
+  for (const k of iter) {
+    m.set(k, getValue(k));
+  }
+  return m;
+};
+
+Array.fromIterable = (iter, getValue) => {
+  const a = [];
+  for (const k of iter) {
+    a.push(getValue(k));
+  }
+  return a;
+};
+
 export default class Imperial {
-  static fromLog(log) {
-    let game = new Imperial();
+  static fromLog(log, initialState) {
+    let game = new Imperial(initialState);
     log.forEach((entry) => game.tick(entry));
     return game;
   }
 
-  constructor() {
+  constructor(initialState) {
+    this.players = {};
     this.availableActions = new Set();
     this.log = [];
-    this.nations = this.setupNations();
-    this.players = {};
-    this.provinces = this.setupProvinces();
-    this.rondelSlots = this.setupRondelSlots();
+
+    this.nations = Map.fromIterable(Nation, () => ({
+      controller: initialState === undefined ? "" : null,
+      treasury: initialState === undefined ? 0 : 11,
+      rondelPosition: initialState === undefined ? "" : null,
+    }));
+    this.provinces = Map.fromIterable(Strings.Province, () => ({
+      unitCount: 0,
+      hasFactory: true,
+    }));
+    this.rondelSlots = Array.fromIterable(Strings.Rondel, (name) => ({
+      name,
+      nations: [],
+    }));
+
+    if (initialState !== undefined) {
+      this.order = initialState.players.map((p) => p.id);
+      for (const { id, nation } of initialState.players) {
+        this.players[id] = {
+          name: id,
+          bonds: [
+            { nation, cost: 9 },
+            { nation: NationCard.get(nation), cost: 2 },
```

Comment:
I'm not sure how I feel about `get`ting a nation off a `NationCard` as being the 2-cost bond one but I can't think of a better alternative so I guess it's fine. Alternatives I can think of suffer from being more verbose/holding more data.

I do wonder how you feel about making constants out of `Bond`s. Do you think that makes sense? I think memoizing them could make sense.